### PR TITLE
ED-2142 case study input data validation

### PR DESCRIPTION
**Pull request by kowalcj0**
This [ticket](https://uktrade.atlassian.net/browse/ED-2142) & this [bug ticket](https://uktrade.atlassian.net/browse/ED-2248)

Scenario:
```gherkin

  @ED-2142
  @fab
  @case-study
  @profile
  Scenario: Supplier should not be able to use invalid values when adding a case study
    Given "Peter Alder" created an unverified profile for randomly selected company "Y"

    When "Peter Alder" attempts to add a case study using following values
      |field         |value type      |separator |error                                                                                |
      |title         |61 characters   |          |Ensure this value has at most 60 characters (it has 61).                             |
      |title         |empty string    |          |This field is required.                                                              |
      |summary       |201 characters  |          |Ensure this value has at most 200 characters (it has 201).                           |
      |summary       |empty string    |          |This field is required.                                                              |
      |description   |1001 characters |          |Ensure this value has at most 1000 characters (it has 1001).                         |
      |description   |empty string    |          |This field is required.                                                              |
      |sector        |invalid sector  |          |Select a valid choice. this is an invalid sector is not one of the available choices.|
      |website       |256 characters  |          |Enter a valid URL.                                                                   |
      |keywords      |book, keys, food|pipe      |You can only enter letters, numbers and commas.                                      |
      |keywords      |book, keys, food|semi-colon|You can only enter letters, numbers and commas.                                      |
      |keywords      |book, keys, food|colon     |You can only enter letters, numbers and commas.                                      |
      |keywords      |book, keys, food|full stop |You can only enter letters, numbers and commas.                                      |
      |keywords      |empty string    |          |This field is required.                                                              |
      |image_1       |invalid image   |          |Invalid image format, allowed formats: PNG, JPG, JPEG                                |
      |image_1       |no image        |          |This field is required.                                                              |
      |caption_1     |121 characters  |          |Ensure this value has at most 120 characters (it has 121).                           |
      |caption_1     |empty string    |          |This field is required.                                                              |
      |image_2       |invalid image   |          |Invalid image format, allowed formats: PNG, JPG, JPEG                                |
      |caption_2     |121 characters  |          |Ensure this value has at most 120 characters (it has 121).                           |
      |image_3       |invalid image   |          |Invalid image format, allowed formats: PNG, JPG, JPEG                                |
      |caption_3     |121 characters  |          |Ensure this value has at most 120 characters (it has 121).                           |
      |testimonial   |1001 characters |          |Ensure this value has at most 1000 characters (it has 1001).                         |
      |source_name   |256 characters  |          |Ensure this value has at most 255 characters (it has 256).                           |
      |source_job    |256 characters  |          |Ensure this value has at most 255 characters (it has 256).                           |
      |source_company|256 characters  |          |Ensure this value has at most 255 characters (it has 256).                           |

    Then "Peter Alder" should see expected case study error message


  @ED-2142
  @fab
  @case-study
  @profile
  @bug
  @ED-2248
  @fixme
  Scenario: Supplier should not be able to use invalid values when adding a case study
    Given "Peter Alder" created an unverified profile for randomly selected company "Y"

    When "Peter Alder" attempts to add a case study using following values
      |field         |value type      |separator |error             |
      |website       |invalid http    |          |Enter a valid URL.|
      |website       |invalid https   |          |Enter a valid URL.|

    Then "Peter Alder" should see expected case study error message
```

### CODE REVIEW

File: tests/functional/features/pages/fab_ui_case_study_images.py [hunk 1]
```diff
@@ -66,8 +62,31 @@ def read_image(file_path: str):
...
+
+    def get_mimetype(path):
+        if path is not None:
+            res = MimeTypes().guess_type(path)[0]
+        else:
+            res = ""
+        return res
+
+    def read_image(file_path: str):
+        if file_path is not None:
```
Comment (konradko) [inline tests/functional/features/pages/fab_ui_case_study_images.py]: you probably want `if os.path.exists(path)`

File: tests/functional/features/steps/fab_when_impl.py [hunk 2]
```diff
@@ -1623,3 +1631,90 @@ def fab_submit_verification_code(context, supplier_alias):
...
+
+    results = []
+    for row in table:
+        case_study = random_case_study_data("test")
+        field = row["field"]
+        value_type = row["value type"]
+        separator = row["separator"]
+        error = row["error"]
+
+        if value_type.endswith(" characters"):
```
Comment (konradko) [inline tests/functional/features/steps/fab_when_impl.py]: For readability, I'd turn this huge if:else into a dictionary

File: tests/functional/features/pages/fab_ui_case_study_images.py (hunk 1 above)
Comment (kowalcj0) [reply tests/functional/features/pages/fab_ui_case_study_images.py]: @konradko This makes sense but there's a use case where we don't want to send a required image (1st case study image) then we set `path` value to `None`. Using `os.path.exists` would then fail.

Comment (kowalcj0) [reply tests/functional/features/pages/fab_ui_case_study_images.py]: PS. I've refactored those local functions, so they're a bit simpler

File: tests/functional/features/steps/fab_when_impl.py (hunk 2 above)
Comment (kowalcj0) [reply tests/functional/features/steps/fab_when_impl.py]: @konradko I moved this huge `IF` to a dedicated function.

**Pull request merged**